### ISSUE TITLE
WOQL idgen is outdated and should be brought in line with the document interface id generation

### ISSUE THREAD

**Issue by matko** [terminusdb/terminusdb#1098]
WOQL's `idgen` (and probably `random_idgen` too) generates IDs that make little sense given the current id generation rules introduced with the document interface. We should rethink this word and make it fit more properly.

I think `idgen` should be of the format `idgen(+Type, +ID_Components, -ID)`, where `Type` is the type for which we generate the id, `ID_Components` is a list of strings containing the suffix parts, and `ID` is the generated ID. For example, `idgen("Employee", ["Steve"], ID)` should unify `ID` with `Employee/Steve`. `idgen("Employee",["Steve","work_day","monday"], ID)` should unify `ID` with `Employee/Steve/work_day/monday`. Each suffix part could potentially be a var.

This should be sufficient for allowing people to generate valid IDs for types that have either no key strategy or the random strategy. With knowledge of how the key generation strategy works, it may also allow them to generate proper IDs for lexical types. However, we should probably encourage people to use the woql words for document inserts rather than raw triple inserts, as it'll be able to do the id generation far better.

Alternatively, we could consider something like `idgen({"@type": "Employee", "name": "Steve"}, ID)`, where the given document is used purely for id generation.

**Comment by spl** [comment]
We plan to deprecate `idgen` in WOQL and eventually remove it. You can use `insert_document`/`update_document` from WOQL now. In the future, we want to expose patch to replace uses of WOQL for inserting/updating.

**Comment by wcoolidge** [comment]
My vote is for the consistency of doc with triple primitives rather than deprecating idgen. I build up documents from add_triple and rely on idgen and would be sorry to see this feature go in deference to working on a doc only basis. I found this issue as I have a bug in this area (Not well formed WOQL JSON-LD) which appears to be a regression.

**Comment by hoijnet** [comment]
I agree with @wcoolidge. I find strength of WOQL and TDB to be the very interop between docs and triples, leveraging the notion of having more than one way do to things. Deprecating these low-level mechanics will remove some of the beauty of the platform IMHO and detract from the value probably. 

Having the semantics of uuid/guids as I'd generation targets would be helpful for me (including the different kinds of uuids) for when almost truly random ids are useful, and if needed context/product bound. 

For some things, a fixed ID may have issues, as the id may need to change as the reference is no longer the same, but the referenced thing is still mostly the same (consider people changing their name). There is both a benefit and an issue with bound id's (and this topic is probably the vi vs emacs camps I think 😉 and each know their own modelling preference).

**Comment by hoijnet** [comment]
Adding, exposing the id generation for the random id could be a good middle ground, to leverage only one dependable implementation perhaps?

**Comment by GavinMendelGleason** [comment]
Indeed, random ids should be trivial. Lexical and Hash can sometimes be done without too much trouble. ValueHash is very hard.

**Comment by matko** [comment]
I understand the wish to keep the interface for adding raw triples, but I'd like to give our developer perspective on this issue here.

Currently, the document interface is a three stage process.
- In the first stage, we analyze the input json using the type information in the schema. We figure out the proper types for each field, and we ensure that all data is of the right type, and you didn't make any cardinality mistakes.
- The second stage generates all the triples and inserts them into a new layer, building up a new commit.
- Finally, the new commit is checked against the schema. This does a lot of the type checking that we actually already did in the first stage (data type checking, cardinality checking, etc) and additionally does a lot of checks for things that could go wrong when you directly insert triples rather than go through the document interface (building your lists wrong, giving your objects an ID outside of its assigned namespace, etc), and it checks referential integrity.

In our efforts to speed up document interface speed, we've come to realize that this duplicate work in the schema checker has to go. Given that you're inserting or modifying through the document interface, there's a lot of checks at schema checking time that we simply do not have to do. The only thing that still needs to be done at check time is checking referential integrity.

So in theory we can do schema checking much faster. However, as we're also supporting direct triple manipulation, we're forced to stick with the extended version of schema checking. For future development, we have to consider one of the following options:
- Keep supporting direct triple manipulation, and therefore keep supporting the old schema checker which does very extensive checks, while having a more minimal version of the schema checker which we run when someone only did modifications through trusted interfaces like the document interface.
- Scrap direct triple manipulation, and have only one version of the schema checker.

Supporting multiple schema checking pathways is a burden on development. Already we're occasionally running into issues where direct triple manipulation lets you insert wrong data, stuff we never caught in our own tests cause the document interface doesn't let you construct such situations. If we wish to continue supporting direct triple manipulation, we essentially have to double our tests, where whatever we test through the document interface, we also test through direct triple inserts, to make sure the behavior remains the same. Furthermore, we really need to expand on the test set for direct triple manipulation, knowing that this completely avoids the safety net of the trusted interfaces and could insert any arbitrary garbage.

The question is if that's really worth it. Any development going into continued support of direct triple manipulation is development effort that cannot go elsewhere. In my opinion, the time could really be spent much better elsewhere.

**Comment by hoijnet** [comment]
Great summary and understand the issue much better considering the effort needed. 

It's a thorny one, and ultimately depend on the product vision of what TerminusDB is and should/will be. I see it as the perfect blend of a document and triples database, which is an extremely powerful combination. Referential integrity is of utmost importance as you say, and performance is key. Doubling the implementation will lead to very hard to find errors.

This is a very tough one indeed. To me, with a brief thinking session on it, it's hard to say which is a good route ahead. At the moment I use mostly document manipulation and it is a great way to interfact. However, making manipulations at scale across the database using WOQL in the future would likely be hard without the ability to manipulate individual triples, and TerminusDB is so much more powerful with WOQL than a document database with schema in my mind.

Unsure if triples manipulation is the key or not though, or if it makes sense to live without it for the reasons mentioned.

---
As a practical note, `idgen` is necessary to generate correct ids for example in `insert_document` with generated JSON documents to submit. Such as through importing by CSV, so a good solution is needed there depending on ID generation needs. Perhaps would need a different interface if `idgen` is deprecated. I don't know if it's possible to `typecast` to id via typecast or if it has to be done through `idgen`?

**Comment by hoijnet** [comment]
I'm probably just afraid to  lose some power in the WOQL interface, you will make a good decision! Perhaps there is a third way? Some of the WOQL usecases I can envision will become much harder is how to manipulate already-created subdocuments; as complex documents would have to be manipulated one by one. Or, the editing would have to be done with the schema checker off; which would be dangerous as the database could get broken without realizing it.

Not easy! Completely agree!

**Comment by wcoolidge** [comment]
Your points on type safety, performance, maintenance effort are all well taken. For those that are fond of WOQL style of working with triples, consistency going forward would be important. Wondering if there an option of exposing or documenting the underlying primitives (for the consistency) that you are using (or generating to) to define the triples in the second stage so that there is no redundancy or inconsistency ?  I would assume that these users would have to take on type responsibility to some degree.  In other words, is documenting an underlying API an option/solution?

**Comment by hoijnet** [comment]
Some more thoughts to share this morning on how to add triples to lists/sets from capture operations would need some very interesting in-document triples/data-nuggets manipulations  (delete/addToList, setSingleProperty, etc). Manipulating (sub)documents that way would be very intuitive, but also take quite some time to develop to cover all use cases; yet likely quite worth it.

It amasses to a different way of thinking, but also likely a better end result and I really like the perspective. Subdocuments would require some kind of universal path notation for the manipulation selection, but such exist.

I continue to be torn short-term 😄

**Comment by GavinMendelGleason** [comment]
Regarding updates, we're very soon going to have a way to perform patches to documents, which would make it fairly easy to do updates to long paths along a document. This should make it easier to do what is currently done with add/remove triples without having to do the id management at all.

**Comment by hoijnet** [comment]
Cool! Looking forward to understanding more 🤩

**Comment by GavinMendelGleason** [comment]
Ok, I think we're going to go with two schema checkers. If triples updates are made, then you'll have a slightly slower schema check. And we'll try to make idgen more compatible with the appropriate IDs that would be generated from the doc interface. This will not allow you to do ValueHash, but other than that might work decently.

**Comment by hoijnet** [comment]
I like the approach taken, thanks a lot for taking all perspectives into consideration. Agree in the approach prioritising performance for the documents interface over the triples one! No an easy decision you have made. I believe it aligns with what you are trying to accomplish overall! It will be great to see how the different parts come together ⚡️

**Comment by dfrnt-SigmundFuchs** [comment]
This issue has been closed due to inactivity. Please don’t hesitate to reopen it if you have any further questions or would like to continue the discussion.
